{
 "metadata": {
  "name": "",
  "signature": "sha256:11538dcc55a76ef7dcb8e38e8dc95cf194ceae66f9684e0d439b35ed0522f694"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Imports\n",
      "from study_1001_networks import open_saved\n",
      "import numpy as np\n",
      "import pandas as pd\n",
      "\n",
      "(path, data, raw_data, edge_list, node_list, meta_data, connected,\n",
      "full_data, filtered_edges) = open_saved()\n",
      "\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Load files\n",
      "f = open('/Users/jimbo/qiime/study_1001/output/dn_otu/rep_set/study_1001_seqs_rep_set.fasta', 'rb')\n",
      "sequences = f.read()\n",
      "sequences = sequences.split('\\n')\n",
      "f.close()\n",
      "\n",
      "\n",
      "# Extract sequence data I need\n",
      "seq_IDs = [ s.split(' ')[0][1:] for s in sequences[::2]]  # Takes just the \"denovo#\" section of the sequence names (even rows)\n",
      "seq_strings = [s for s in sequences[1::2]]\n",
      "seq_IDs = seq_IDs[:-1]\n",
      "\n",
      "sequences = dict(zip(seq_IDs, seq_strings))  # Dict of format{ denovo#: 'AAAAT...'}\n",
      "\n",
      "\n",
      "# Extract name data I need.\n",
      "g = open('/Users/jimbo/qiime/study_1001/output/dn_otu/uclust_assigned_taxonomy/study_1001_seqs_rep_set_tax_assignments.txt','rb')\n",
      "seq_names = g.read()\n",
      "seq_names = seq_names.split('\\n')\n",
      "g.close()\n",
      "\n",
      "# split at tabs and remove any rows that do not conform to data pattern\n",
      "seq_name_split = [s.split('\\t') for s in seq_names]\n",
      "seq_name_split = [x for x in seq_name_split if len(x) >1]    \n",
      "\n",
      "# Get phyla strings and strip species section. \n",
      "phyla = [ row[1] for row in seq_name_split ]\n",
      "phyla = [string.split('; s')[0] for string in phyla ]\n",
      "\n",
      "\n",
      "# Create dictionary of format {denovo# : 'k__p__c__...'}\n",
      "names = [ row[0] for row in seq_name_split ]\n",
      "phyla_OTU_IDs = dict(zip(names, phyla))  \n",
      "\n",
      "#print len(phyla_OTU_IDs)\n",
      "#print len(phyla_OTU_IDs.values())\n",
      "\n",
      "#from collections import Counter\n",
      "#Counter(phyla_OTU_IDs.values())\n",
      "phyla_set = set(phyla_OTU_IDs.values())\n",
      "#print len(phyla_set)\n",
      "\n",
      "def method1(a_dict,a_value):\n",
      "    for key,value in a_dict.iteritems():\n",
      "        if value == a_value:\n",
      "            return key\n",
      "\n",
      "shortened_dict = {}\n",
      "for taxa in phyla_set:\n",
      "    if taxa not in shortened_dict:\n",
      "        shortened_dict[taxa] = method1(phyla_OTU_IDs,taxa)\n",
      "\n",
      "print len(shortened_dict)\n",
      "\n",
      "\n",
      "# We have 1068 unique OTUs in these samples. Each OTU 16s sequence is >97% similar to other sequences\n",
      "# easiest thing is to get a dataframe with a row for each of 1000 unique values and a seq_ID\n",
      "# then add the sequence later.\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "1068\n"
       ]
      }
     ],
     "prompt_number": 12
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Let's refactor all of the shitty ways I'm doing taxa string splitting\n",
      "#and just have one way to do it that makes esnse\n",
      "\n",
      "df = pd.DataFrame({'taxa' : shortened_dict.keys(), 'seq_IDs' : shortened_dict.values() })\n",
      "df['sequence'] == na\n",
      "\n",
      "print type(sequences)\n",
      "print len(sequences)\n",
      "\n",
      "for key, val in sequences.iteritems():\n",
      "    df.loc[df.seq_IDs==key, 'sequence'] = val\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "def taxa_string_split(df, name_column, columns = ['name', \"kingdom\", \"phylum\", \"class\", \n",
      "                                                  \"order\", \"family\",\"genus\"], depth=6):\n",
      "    \n",
      "    \"\"\" This assumes you have a dataframe with one (or more) name columns. \n",
      "        By default it splits a taxa string in the QIIME format:\n",
      "        \"k__Bacteria; p__Verrucomicrobia; c__Opitutae; o__Opitutales; f__Opitutaceae;\n",
      "        g__Opitutus; s__\"\n",
      "        into individual columns to make comparisons at many different phylogenetic levels possible.\n",
      "        It accepts different column names and taxonomic naming depths. \"\"\"\n",
      "    \n",
      "    # I want to add columns in place to existing dataframe by splitting strings and assigning\n",
      "    # positions in the resulting list to new columns\n",
      "    \n",
      "    #df[columns[1]] = df[name_column]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "1068\n",
        "<type 'dict'>\n",
        "63135\n",
        "          seq_IDs                                               taxa  \\\n",
        "0      denovo4524  k__Bacteria; p__Bacteroidetes; c__Cytophagia; ...   \n",
        "1      denovo1777  k__Bacteria; p__Firmicutes; c__Bacilli; o__Bac...   \n",
        "2     denovo43640  k__Bacteria; p__Proteobacteria; c__Betaproteob...   \n",
        "3     denovo23997  k__Bacteria; p__Proteobacteria; c__Deltaproteo...   \n",
        "4     denovo58852  k__Bacteria; p__Bacteroidetes; c__VC2_1_Bac22;...   \n",
        "5     denovo50178  k__Bacteria; p__Proteobacteria; c__Betaproteob...   \n",
        "6     denovo54422  k__Bacteria; p__Firmicutes; c__Clostridia; o__...   \n",
        "7     denovo33307  k__Bacteria; p__Firmicutes; c__Clostridia; o__...   \n",
        "8     denovo26974  k__Bacteria; p__Actinobacteria; c__Actinobacte...   \n",
        "9      denovo9618  k__Bacteria; p__Proteobacteria; c__Betaproteob...   \n",
        "10    denovo49308  k__Bacteria; p__Proteobacteria; c__Betaproteob...   \n",
        "11    denovo58974  k__Bacteria; p__Proteobacteria; c__Alphaproteo...   \n",
        "12    denovo36384  k__Bacteria; p__Bacteroidetes; c__Cytophagia; ...   \n",
        "13     denovo4529  k__Bacteria; p__Firmicutes; c__Clostridia; o__...   \n",
        "14    denovo34481  k__Bacteria; p__Proteobacteria; c__Alphaproteo...   \n",
        "15    denovo28361  k__Bacteria; p__Bacteroidetes; c__[Saprospirae...   \n",
        "16    denovo26849  k__Bacteria; p__Actinobacteria; c__Actinobacte...   \n",
        "17    denovo20041  k__Bacteria; p__Actinobacteria; c__Actinobacteria   \n",
        "18    denovo29463  k__Bacteria; p__Bacteroidetes; c__Flavobacteri...   \n",
        "19    denovo49893         k__Bacteria; p__TM6; c__F38; o__; f__; g__   \n",
        "20    denovo16795  k__Bacteria; p__Firmicutes; c__Clostridia; o__...   \n",
        "21    denovo14880  k__Bacteria; p__Cyanobacteria; c__4C0d-2; o__M...   \n",
        "22    denovo31171  k__Bacteria; p__Proteobacteria; c__Alphaproteo...   \n",
        "23    denovo28423  k__Bacteria; p__Bacteroidetes; c__Cytophagia; ...   \n",
        "24    denovo12990  k__Bacteria; p__Actinobacteria; c__Actinobacte...   \n",
        "25     denovo8457  k__Bacteria; p__Chlamydiae; c__Chlamydiia; o__...   \n",
        "26     denovo6124  k__Bacteria; p__Proteobacteria; c__Gammaproteo...   \n",
        "27    denovo24317  k__Bacteria; p__Bacteroidetes; c__Cytophagia; ...   \n",
        "28    denovo34488  k__Bacteria; p__Proteobacteria; c__Alphaproteo...   \n",
        "29     denovo7630  k__Bacteria; p__Bacteroidetes; c__Flavobacteri...   \n",
        "...           ...                                                ...   \n",
        "1038   denovo6007  k__Bacteria; p__Proteobacteria; c__Gammaproteo...   \n",
        "1039  denovo20535  k__Bacteria; p__Firmicutes; c__Bacilli; o__Bac...   \n",
        "1040  denovo62248  k__Bacteria; p__Acidobacteria; c__DA052; o__El...   \n",
        "1041  denovo32613  k__Bacteria; p__Actinobacteria; c__Actinobacte...   \n",
        "1042  denovo11063  k__Bacteria; p__Proteobacteria; c__Gammaproteo...   \n",
        "1043  denovo18241  k__Bacteria; p__Proteobacteria; c__Gammaproteo...   \n",
        "1044  denovo36887  k__Bacteria; p__Actinobacteria; c__Actinobacte...   \n",
        "1045  denovo34774  k__Bacteria; p__Firmicutes; c__Clostridia; o__...   \n",
        "1046  denovo56972  k__Bacteria; p__Actinobacteria; c__Actinobacte...   \n",
        "1047  denovo11746  k__Bacteria; p__Cyanobacteria; c__4C0d-2; o__Y...   \n",
        "1048  denovo36736  k__Bacteria; p__Proteobacteria; c__Gammaproteo...   \n",
        "1049  denovo10328  k__Bacteria; p__Actinobacteria; c__Actinobacte...   \n",
        "1050  denovo21447  k__Bacteria; p__Proteobacteria; c__Alphaproteo...   \n",
        "1051  denovo30028  k__Bacteria; p__Proteobacteria; c__Betaproteob...   \n",
        "1052  denovo60880  k__Bacteria; p__SBR1093; c__A712011; o__; f__;...   \n",
        "1053  denovo36739  k__Bacteria; p__Proteobacteria; c__Alphaproteo...   \n",
        "1054  denovo42243  k__Bacteria; p__Actinobacteria; c__Coriobacter...   \n",
        "1055  denovo44468  k__Bacteria; p__Planctomycetes; c__BD7-11; o__...   \n",
        "1056  denovo32260  k__Bacteria; p__Chloroflexi; c__Chloroflexi; o...   \n",
        "1057  denovo47339  k__Bacteria; p__Verrucomicrobia; c__Verrucomic...   \n",
        "1058  denovo21192  k__Bacteria; p__Proteobacteria; c__Betaproteob...   \n",
        "1059  denovo21815  k__Bacteria; p__Actinobacteria; c__Actinobacte...   \n",
        "1060   denovo7642      k__Bacteria; p__OP3; c__koll11; o__; f__; g__   \n",
        "1061  denovo20531  k__Bacteria; p__Proteobacteria; c__Alphaproteo...   \n",
        "1062  denovo34489  k__Bacteria; p__Actinobacteria; c__Actinobacte...   \n",
        "1063  denovo11828  k__Bacteria; p__Proteobacteria; c__Deltaproteo...   \n",
        "1064  denovo29175  k__Bacteria; p__Actinobacteria; c__Acidimicrob...   \n",
        "1065  denovo31354  k__Bacteria; p__Proteobacteria; c__Gammaproteo...   \n",
        "1066  denovo47016  k__Bacteria; p__Proteobacteria; c__Deltaproteo...   \n",
        "1067  denovo60753     k__Bacteria; p__Chloroflexi; c__Thermomicrobia   \n",
        "\n",
        "                                               sequence  \n",
        "0     TACGTAGGTGGCAAGCGTTGTCCGGATTTATTGGGTTTAAAGGGTG...  \n",
        "1     TACGGAGGGGGCTAGCGTTGTTCGGAATTACTGGGCGTAAAGCGCG...  \n",
        "2     TACGTAGGGTGCGAGCGTTAATCGGAATTACTGGGCGTAAAGGGTG...  \n",
        "3     TACGTAGGGTGCAAGCGTTGTTCGGAATCACTGGGCGTAAAGGGCG...  \n",
        "4     TACGGAGGATGCAAGCGTTATCCGGATTCATTGGGTTTAAAGGGTG...  \n",
        "5     TACGTAGGGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGTG...  \n",
        "6     TACGTAGGGTGCGAGCGTTATTCGGAATTACTGGGCGTAAAGGGTG...  \n",
        "7     TACGTAGGGGGCGAGCGTTGTCCGGAATGACTGGGCGTAAAGGGTA...  \n",
        "8     TACGTAGGGTGCAAGCGTTGTCCGGAATTATTGGGCGTAAAGAGCT...  \n",
        "9     TACGTAGGGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGGGTG...  \n",
        "10    TACGAAGGGGGCTAGCGTTGCTCGGAATTACTGGGCGTAAAGCGCA...  \n",
        "11    TACGGAGGGGGCTAGCGTTGTTCGGAATTACTGGGCGTAAAGCGCA...  \n",
        "12    TACGGAGGGTGCGAGCGTTGTCCGGATTTATTGGGTTTAAAGGGTG...  \n",
        "13    TACGTAGGGGGCAAGCGTTATCCGGATTTACTGGGTGTAAAGGGAG...  \n",
        "14    TACGAAGGGGGCTAGCGTTGTTCGGAATTACTGGGCGTAAAGCGCG...  \n",
        "15    TACGGAGGGTGCAAGCGTTATCCGGATTCACTGGGTTTAAAGGGAG...  \n",
        "16    TACGGAGGTGGCAAGCGTTGTCCGGATTTATTGGGCGTAAAGAGCT...  \n",
        "17    TACGAAGGGGGCTAGCGTTGCTCGGAATTACTGGGCGTAAAGAGCT...  \n",
        "18    TACGGAGGATCCAAGCGTTATCCGGAATCATTGGGTTTAAAGGGTC...  \n",
        "19    TACGAGAGGTGCAAGCGTTATTCGGAATTATTGGGCGTAAAGGGCG...  \n",
        "20    TACGTAGGTGGCGAGCGTTGTCCGGAATTACTGGGTGTAAAGGGTG...  \n",
        "21    TACGTAGGTGGCAAGCGTTGTCCGGATTTATTGGGCGTAAAGAGTT...  \n",
        "22    TACGAAAGGGGCAGGCGTTGTTCGGATTTACTGGGCGTAAAGGGCA...  \n",
        "23    TACGTAGGTGGCAAGCGTTGTCCGGATTTATTGGGTTTAAAGGGTG...  \n",
        "24    TACGTAGGGTGCAAGCGTTATCCGGAATTATTGGGCGTAAAGAGCT...  \n",
        "25    TACGGAGGGTGCAAGCATTAATCGGATTTATTGGGTGTAAAGGGCA...  \n",
        "26    TACGGAGGGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCA...  \n",
        "27    TACGGAGGGTGCAAGCGTTGTCCGGATTTATTGGGTTTAAAGGGTG...  \n",
        "28    TACGAAGGGGGCTAGCGTTGTTCGGATTTACTGGGCGTAAAGCGCA...  \n",
        "29    TACGGAGGGGGCAAGCGTTGTTCGGAATCATTGGGCTTAAAGGGTG...  \n",
        "...                                                 ...  \n",
        "1038  TACGGAGGGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCG...  \n",
        "1039  TACGTAGGGGGCGAGCGTTGTCCGGAATTATTGGGCGTAAAGCGCG...  \n",
        "1040  TACGAAGGGTGCAAGCGTTACTCGGAATTACTGGGCGTAAAGGGTT...  \n",
        "1041  TACGTAGGGCGCAAGCGTTGTCCGGAATTATTGGGCGTAAAGAGCT...  \n",
        "1042  TACGGAGGGTGCGAGCGTTAATCGGAATTACTGGGCGTAAAGCGCA...  \n",
        "1043  TACAGAGGGTGCAAGCGTTAATCGGAATAATTGGGCGTAAAGGGTG...  \n",
        "1044  TACGTAGGGTACGAGCGTTGTCCGGAATTATTGGGCGTAAAGAGCT...  \n",
        "1045  TACGTAGGGGGCAAGCGTTATCCGGAATTACTGGGTGTAAAGGGAG...  \n",
        "1046  TACAGAGGGTGCGAGCGTTGTCCGGAATTATTGGGCGTAAAGGGCT...  \n",
        "1047  GACGTAGGATGCAAGCGTTGTCCGGAATTATTGGGCGTAAAGAGTT...  \n",
        "1048  TACAGAGACTGCGAGCGTTATTCGGATTCACTGGGCGTAAAGCGTG...  \n",
        "1049  TACGTAGGGCGCGAGCGTTGTCCGGAATTATTGGGCGTAAAGAGCT...  \n",
        "1050  TACGAAGGGGGCTAGCGTTGTTCGGAATCACTGGGCGTAAAGCGTA...  \n",
        "1051  TACGGAGGGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGTG...  \n",
        "1052  TACGGAGGGAGCAAACGTTACTCGGATTTACTGGGCGTAAAGGGCT...  \n",
        "1053  TACGGAGGGGGCTAGCGTTGTTCGGAATCACTGGGCGTAAAGCGCA...  \n",
        "1054  TACGTAGGGAGCAAGCGTTATCCGGATTTATTGGGCGTAAAGCGCG...  \n",
        "1055  GACGTAGGTGGCAAGCGTTGTTCGGAATCACTGGGCATAAAGCGCG...  \n",
        "1056  GACGGAGGGGGCAAGCGTTGTCCGGAATCACTGGGCGTAAAGCGCG...  \n",
        "1057  CACGGAGGCCCCGAGCGTTGTTCGGAATTACTGGGCGTAAAGAGTG...  \n",
        "1058  TACGTAGGGTCCGAGCGTTGTCCGGAATTACTGGGCGTAAAGCGTG...  \n",
        "1059  GACAGAGGGTGCAAACGTTGCTCGGAATTACTGGGCGTAAAGCGCG...  \n",
        "1060  TACAGAGGCGGCAAGCGTTGTTCGGATTTATTGGGTGCAAAGGGCA...  \n",
        "1061  TACGAAGGGGGCTAGCGTTGCTCGGAATTACTGGGCGTAAAGCGCG...  \n",
        "1062  TACGTAGGGCGCAAGCGTTGTCCGGAATTATTGGGCGTAAAGAGCT...  \n",
        "1063  TACGTAGGGGGCGAGCGTTGTCCGGAATTATTGGGCGTAAAGCGTG...  \n",
        "1064  GACGTAGGGAGCAAGCGTTGTCCGGATTTATTGGGCGTAAAGAGCT...  \n",
        "1065  TACGGATGGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCG...  \n",
        "1066  TACGTAGGGTGCGAGCGTTGTCCGGAATTATTGGGCGTAAAGCGCG...  \n",
        "1067  TACGAAGGGGGCTAGCGTTGTCCGGAGTTACTGGGCGTAAAGGGCG...  \n",
        "\n",
        "[1068 rows x 3 columns]"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n"
       ]
      }
     ],
     "prompt_number": 15
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "I need to convert the above data into arrays or dictionaries. Then I need to compute hamming/levenschtein distances for each pair of sequences. First I should a) get hamming distance working, b) eliminate sequences with no interaction or not enough data to compute interaction. "
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "df.to_pickle('/Users/jimbo/Desktop/working/study_1001_results/rep_otu_sequences')\n",
      "\n",
      "\n",
      "def hamming_distance(s1, s2):\n",
      "    #Return the Hamming distance between equal-length sequences\n",
      "    if len(s1) != len(s2):\n",
      "        raise ValueError(\"Undefined for sequences of unequal length\")\n",
      "    return sum(ch1 != ch2 for ch1, ch2 in zip(s1, s2))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 17
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Get OTU names of interest\n",
      "edge_names = edge_list['feature1'].tolist()\n",
      "edge_names2 = edge_list['feature2'].tolist()\n",
      "edge_names = edge_names + edge_names2\n",
      "names_set = set(edge_names)\n",
      "\n",
      "print len(names_set)\n",
      "\n",
      "sequences_subset = {}\n",
      "phyla_OTU_IDs_subset = {}\n",
      "values = []\n",
      "\n",
      "i=0\n",
      "missing_val = []\n",
      "\n",
      "for key,value in phyla_OTU_IDs.iteritems():\n",
      "    valuestrip = value.replace(' ', '')\n",
      "    if valuestrip in names_set:\n",
      "        if valuestrip not in values:\n",
      "            values.append(valuestrip)\n",
      "            phyla_OTU_IDs_subset[key] = valuestrip\n",
      "    elif valuestrip not in names_set:\n",
      "        if valuestrip not in missing_val:\n",
      "            missing_val.append(valuestrip)\n",
      "            #print valuestrip\n",
      "            i+=1\n",
      "        #pass\n",
      "\n",
      "#print missing_val\n",
      "#print len(missing_val)\n",
      "#print i\n",
      "print phyla_OTU_IDs_subset\n",
      "\n",
      "# There are 171 names in names_set, \n",
      "# a list of names that came from taking all unique taxa strings from both columns of \n",
      "# edge list, a list of all non-NAN rows in the original ccrepe R output. \n",
      "# There are 80 or 91 missing names in phyla_OTU_IDs_subset which was supposed to just\n",
      "\n",
      "\n",
      "            \n",
      "i=0\n",
      "for name in names_set:\n",
      "    if name not in phyla_OTU_IDs_subset.values():\n",
      "        i+\n",
      "        #print name\n",
      "        \n",
      "#print i\n",
      "#print len(phyla_OTU_IDs_subset)\n",
      "#print len(sequences_subset)\n",
      "#print phyla_OTU_IDs_subset\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "171\n",
        "{'denovo62585': 'k__Bacteria;p__Proteobacteria;c__Deltaproteobacteria;o__Syntrophobacterales;f__Syntrophobacteraceae', 'denovo4723': 'k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae', 'denovo12302': 'k__Bacteria;p__Chloroflexi;c__TK17;o__;f__', 'denovo16798': 'k__Bacteria;p__Proteobacteria;c__Deltaproteobacteria;o__Myxococcales;f__Haliangiaceae', 'denovo58441': 'k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Oceanospirillales;f__Alcanivoracaceae', 'denovo57727': 'k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Methylococcales;f__Methylococcaceae', 'denovo28366': 'k__Bacteria;p__Proteobacteria;c__Betaproteobacteria;o__Burkholderiales;f__Comamonadaceae', 'denovo62515': 'k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Oceanospirillales;f__', 'denovo16905': 'k__Bacteria;p__WPS-2;c__;o__;f__', 'denovo16983': 'k__Bacteria;p__Proteobacteria;c__Betaproteobacteria;o__Burkholderiales;f__Oxalobacteraceae', 'denovo35907': 'k__Bacteria;p__Firmicutes;c__Bacilli;o__Bacillales;f__Alicyclobacillaceae', 'denovo675': 'k__Bacteria;p__Proteobacteria;c__Betaproteobacteria;o__Burkholderiales;f__', 'denovo50695': 'k__Bacteria;p__Chloroflexi;c__Anaerolineae;o__CFB-26;f__', 'denovo6448': 'k__Bacteria;p__Proteobacteria;c__Deltaproteobacteria;o__Myxococcales;f__Nannocystaceae', 'denovo27595': 'k__Bacteria;p__Acidobacteria;c__Acidobacteria-5;o__;f__', 'denovo1777': 'k__Bacteria;p__Firmicutes;c__Bacilli;o__Bacillales;f__Planococcaceae', 'denovo54533': 'k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Rickettsiales;f__', 'denovo1775': 'k__Bacteria;p__Firmicutes;c__Bacilli;o__Bacillales;f__Paenibacillaceae', 'denovo49308': 'k__Bacteria;p__Proteobacteria;c__Betaproteobacteria;o__Burkholderiales;f__Burkholderiaceae', 'denovo19654': 'k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Rhizobiales;f__Methylocystaceae', 'denovo4529': 'k__Bacteria;p__Firmicutes;c__Clostridia;o__Clostridiales;f__', 'denovo34481': 'k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Rhizobiales;f__Rhodobiaceae', 'denovo31596': 'k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Vibrionales;f__Vibrionaceae', 'denovo36733': 'k__Bacteria;p__Firmicutes;c__Bacilli;o__Bacillales;f__Thermoactinomycetaceae', 'denovo56399': 'k__Bacteria;p__Proteobacteria;c__Betaproteobacteria;o__Methylophilales;f__Methylophilaceae', 'denovo48678': 'k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Rhizobiales;f__Aurantimonadaceae', 'denovo24836': 'k__Bacteria;p__Acidobacteria;c__;o__;f__', 'denovo16795': 'k__Bacteria;p__Firmicutes;c__Clostridia;o__Clostridiales;f__Ruminococcaceae', 'denovo61720': 'k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Rhizobiales;f__', 'denovo36029': 'k__Bacteria;p__Proteobacteria;c__Deltaproteobacteria;o__;f__', 'denovo61583': 'k__Bacteria;p__Proteobacteria;c__Betaproteobacteria;o__;f__', 'denovo43270': 'k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__;f__', 'denovo22824': 'k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Thiotrichales;f__', 'denovo61587': 'k__Bacteria;p__Proteobacteria;c__Betaproteobacteria;o__Burkholderiales;f__Alcaligenaceae', 'denovo12511': 'k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Xanthomonadales;f__Xanthomonadaceae', 'denovo57440': 'k__Bacteria;p__Proteobacteria;c__Betaproteobacteria;o__Methylophilales;f__', 'denovo12513': 'k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Legionellales;f__Legionellaceae', 'denovo42210': 'k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Rhizobiales;f__Brucellaceae', 'denovo29159': 'k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Caulobacterales;f__Caulobacteraceae', 'denovo35595': 'k__Bacteria;p__Proteobacteria;c__Betaproteobacteria;o__Rhodocyclales;f__Rhodocyclaceae', 'denovo12519': 'k__Bacteria;p__Proteobacteria;c__Deltaproteobacteria;o__Myxococcales;f__', 'denovo12518': 'k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Rhizobiales;f__Hyphomicrobiaceae', 'denovo36739': 'k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Rhodobacterales;f__Rhodobacteraceae', 'denovo41479': 'k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Rhizobiales;f__Rhizobiaceae', 'denovo14797': 'k__Bacteria;p__Firmicutes;c__Clostridia;o__Clostridiales;f__Peptococcaceae', 'denovo10797': 'k__Bacteria;p__Firmicutes;c__Clostridia;o__Clostridiales;f__Symbiobacteriaceae', 'denovo41474': 'k__Bacteria;p__Firmicutes;c__Clostridia;o__Clostridiales;f__Clostridiaceae', 'denovo3850': 'k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Rhodobacterales;f__Hyphomonadaceae', 'denovo53825': 'k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Thiotrichales;f__Piscirickettsiaceae', 'denovo34865': 'k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Rhizobiales;f__Bradyrhizobiaceae', 'denovo35426': 'k__Bacteria;p__Acidobacteria;c__Solibacteres;o__Solibacterales;f__Solibacteraceae', 'denovo26335': 'k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Rickettsiales;f__Rickettsiaceae', 'denovo3508': 'k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__', 'denovo27383': 'k__Bacteria;p__Proteobacteria;c__Deltaproteobacteria;o__MIZ46;f__', 'denovo35422': 'k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Erythrobacteraceae', 'denovo35420': 'k__Bacteria;p__Proteobacteria;c__Betaproteobacteria;o__Nitrosomonadales;f__Nitrosomonadaceae', 'denovo18794': 'k__Bacteria;p__Proteobacteria;c__Deltaproteobacteria;o__Myxococcales;f__Myxococcaceae', 'denovo46464': 'k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Rhizobiales;f__Beijerinckiaceae', 'denovo5133': 'k__Bacteria;p__Firmicutes;c__Bacilli;o__Bacillales;f__', 'denovo46783': 'k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Chromatiales;f__', 'denovo56975': 'k__Bacteria;p__Chloroflexi;c__Anaerolineae;o__Caldilineales;f__Caldilineaceae', 'denovo22400': 'k__Bacteria;p__Firmicutes;c__Bacilli;o__Bacillales;f__Staphylococcaceae', 'denovo34774': 'k__Bacteria;p__Firmicutes;c__Clostridia;o__Clostridiales;f__Lachnospiraceae', 'denovo58972': 'k__Bacteria;p__Firmicutes;c__Bacilli;o__Bacillales;f__Bacillaceae', 'denovo40105': 'k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Rhodospirillales;f__Acetobacteraceae', 'denovo38253': 'k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Oceanospirillales;f__Halomonadaceae', 'denovo22216': 'k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Legionellales;f__Coxiellaceae', 'denovo23948': 'k__Bacteria;p__Proteobacteria;c__Deltaproteobacteria;o__Bdellovibrionales;f__Bdellovibrionaceae', 'denovo58976': 'k__Bacteria;p__Proteobacteria;c__Deltaproteobacteria;o__Myxococcales;f__Polyangiaceae', 'denovo39002': 'k__Bacteria;p__Proteobacteria;c__Deltaproteobacteria;o__Bdellovibrionales;f__Bacteriovoracaceae', 'denovo31223': 'k__Archaea;p__Crenarchaeota;c__Thaumarchaeota;o__Nitrososphaerales;f__Nitrososphaeraceae', 'denovo52461': 'k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Rhizobiales;f__Phyllobacteriaceae', 'denovo56393': 'k__Bacteria;p__Planctomycetes;c__Phycisphaerae;o__Phycisphaerales;f__', 'denovo15207': 'k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Pseudomonadales;f__Moraxellaceae', 'denovo23166': 'k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Xanthomonadales;f__Sinobacteraceae', 'denovo36882': 'k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Enterobacteriales;f__Enterobacteriaceae', 'denovo52765': 'k__Bacteria;p__Firmicutes;c__Bacilli;o__Lactobacillales;f__Carnobacteriaceae', 'denovo42218': 'k__Bacteria;p__BRC1;c__PRR-11;o__;f__', 'denovo47339': 'k__Bacteria;p__Verrucomicrobia;c__Verrucomicrobiae;o__Verrucomicrobiales;f__Verrucomicrobiaceae', 'denovo5814': 'k__Bacteria;p__Chloroflexi;c__Thermomicrobia;o__;f__'}"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n"
       ]
      }
     ],
     "prompt_number": 155
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print hamming_distance(sequences[OTU_names[0]], sequences[OTU_names[1]] )"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "16\n"
       ]
      }
     ],
     "prompt_number": 70
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}